object-model/copy-semantics.md

```diff
@@ -0,0 +1,153 @@
+* explain special member functions          [[C++ object model: special member functions]][5]
+
+### Student outcomes
+
+_A list of things "a student should be able to" after the curriculum._
+_The next word should be an action word and testable in an exam._
+_Max 5 items._
+
+A student should be able to:
+
+1. explain what copy semantics accomplish
+  * establishing "equivalent" object state in another object
+2. explain difference between copying a reference and copying a value\*
+3. explain where copies are made
+
+\* In other languages these differences are sometimes referred to as shallow and deep copy.
+
+### Caveats
+
+_This section mentions subtle points to understand, like anything resulting in
+implementation-defined, unspecified, or undefined behavior._
+
+* Compiler-provided (reference based) copies may result in ownership problems (e.g., `char*`).
+
+### Points to cover
+
+_This section lists important details for each point._
+
+* Principle of copying
+  * Copying of trivially copyable types
+  * Copying of non-trivial class types
+  * Copying an object does not change the original
+* Copy Flavors
+  * Copy by value
+  * Copy by reference
+  * No copy
+* Practical applications
+  * Unique_ptr  (has no copy)
+  * Strings (copies the value) 
+
+
+## Advanced: Implementing user-defined copy operations
+
+### Background/Required Knowledge
+
+A student is able to:
+* explain special member functions        [[C++ object model: special member functions]][5]
+
+It helps when a student is able to:
+* use move semantics                      [[C++ object model: move semantics]][4]
+* explain the rule of zero                [[C++ object model: rule-of-zero]][7]
+* explain the rule of five                [[C++ object model: rule-of-five]][6]
+
+### Student outcomes
+
+_A list of things "a student should be able to" after the curriculum._
+_The next word should be an action word and testable in an exam._
+_Max 5 items._
+
+A student should be able to:
+* explain when they have to implement the copy operations for their own type
+  * Copy constructor
+  * Copy assignment operator
+* implement copy operations for their own types
+* _Optional_: explain when copying with basic and strong exception guarantees is useful.
+
+### Caveats
+
+_This section mentions subtle points to understand, like anything resulting in
+implementation-defined, unspecified, or undefined behavior._
+
+* Intricacies when implementing copy operations:
+  * Examples of how _not_ to write copy operations (e.g., `std::auto_ptr`)
+
+### Points to cover
+
+_This section lists important details for each point._
+
+* Copy constructors and copy assignment operators
+  * How compiler generates default copy operations
+  * =default, =delete (No copy)
+  * How-to write your own copy operations
+  * Rule-of-five
+  * Copy assignment operators can be ref-qualified to avoid assigning into temporary objects. (See example #1)
```

Comment:
`example #1`? Where's that?

Reply:
A this was only present in the gdoc during discussion.